### PR TITLE
Handle pattern as editor placeholder

### LINKED ISSUE #2147: Sometimes the placeholder is identified as IdentifierPattern
### Description

Sometimes the placeholder is identified as IdentifierPattern and not as EditorPlaceholderExpr.

### Steps to Reproduce

```
let <#name#> = <#value#>
```
<#name#> //Bug:  is identified as `IdentifierPattern`
<#value#> //Normal: is identified as `EditorPlaceholderExpr`

### LINKED ISSUE #2156: Support parsing of editor placeholders in type positions
### Issue Kind

Other

### Source Code

```swift
let a: <#T#> = x
```

### Description

`<#T#>` currently gets parsed as a `NamedOpaqueReturnTypeSyntax`. We should create a `EditorPlaceholderTypeSyntax` node and parse `<#T#>` as an editor placeholder.

### PR DESCRIPTION
**Pull request by kimdv**
Fixes #2147
Fixes #2156

### CODE REVIEW

File: Sources/SwiftParser/Patterns.swift [hunk 1]
```diff
@@ -70,12 +70,26 @@ extension Parser {
...
-      return RawPatternSyntax(
-        RawIdentifierPatternSyntax(
-          identifier: identifier,
-          arena: self.arena
+      if identifier.tokenText.isEditorPlaceholder {
+        return RawPatternSyntax(
+          RawExpressionPatternSyntax(
+            expression: RawExprSyntax(
+              RawEditorPlaceholderExprSyntax(
+                placeholder: identifier,
```
Comment (kimdv) [inline Sources/SwiftParser/Patterns.swift:78]: The thing here is that the `identifier` includes `<#` and `#>`. 
Not sure how to come around that

Comment (ahoppen) [reply Sources/SwiftParser/Patterns.swift:78]: For now, that’s intentional. I don’t think that splitting the token into `<#`, the contents and `#>` provides too much value for now.

**Review by ahoppen — COMMENTED** [review]
Instead of creating an `ExpressionPatternSyntax`  that contains a `EditorPlaceholderExprSyntax`, I would prefer to introduce a `EditorPlaceholderPatternSyntax` type that we can use here.

File: CodeGeneration/Sources/SyntaxSupport/PatternNodes.swift [hunk 2]
```diff
@@ -182,4 +182,23 @@ public let PATTERN_NODES: [Node] = [
...
+    kind: .editorPlaceholderPattern,
+    base: .pattern,
+    nameForDiagnostics: "editor placeholder",
+    documentation: """
+      An editor placeholder, e.g. `<#pattern#>` that is used in a position that expects a pattern.
+      """,
+    children: [
+      Child(
+        name: "placeholder",
+        deprecatedName: "identifier",
```
Comment (ahoppen) [inline CodeGeneration/Sources/SyntaxSupport/PatternNodes.swift:195]: This is a new node, so it doesn’t have to have a `deprecatedName`. Same for `EditorPlaceholderTypeSyntax`.

File: Tests/SwiftParserTest/TypeTests.swift [hunk 3]
```diff
@@ -216,4 +216,31 @@ final class TypeTests: ParserTestCase {
...
+              )
+            )
+          )
+        ]
+      ),
+      diagnostics: [
+        DiagnosticSpec(message: "editor placeholder in source file")
+      ]
+    )
+  }
```
Comment (ahoppen) [inline Tests/SwiftParserTest/TypeTests.swift:245]: Could you add a test case for an editor placeholder with generic arguments?

File: CodeGeneration/Sources/SyntaxSupport/PatternNodes.swift (hunk 2 above)
Comment (kimdv) [reply CodeGeneration/Sources/SyntaxSupport/PatternNodes.swift:195]: 🤦‍♂️

**Comment by kimdv** [comment]
@swift-ci please test

**Review by ahoppen — APPROVED** [review]

**Pull request merged**